rsocket-core/src/main/java/io/rsocket/core/RSocketClient.java

```diff
@@ -0,0 +1,136 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.rsocket.core;
+
+import io.rsocket.Payload;
+import io.rsocket.RSocket;
+import org.reactivestreams.Publisher;
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+/**
+ * Contract for performing RSocket requests.
+ *
+ * <p>{@link RSocketClient} differs from {@link RSocket} in a number of ways:
+ *
+ * <ul>
+ *   <li>{@code RSocket} represents a "live" connection that is transient and needs to be obtained
+ *       first, typically from a {@code Mono<RSocket>} via {@code flatMap} or block. By contrast,
+ *       {@code RSocketClient} contains a source of connections such as a {@code Mono<RSocket>} and
+ *       transparently obtains a shared connection to use as needed when requests are made. That
+ *       makes {@code RSocketClient} convenient to create once as a singleton and share it across
+ *       multiple places where requests need to be made.
+ *   <li>An application may use {@code RSocket} for sending and it may also implement it for
+ *       receiving. By contrast, {@code RSocketClient} is designed for sending only, and typically
+ *       from the client side where shared connections can be obtained as needed. However it can
+ *       also be {@link #from(RSocket) adapted} for sending through a "live" {@code RSocket} from
+ *       the server side.
+ *   <li>{@code RSocket} does not allow a second subscription for the same request in part because
+ *       there is no way to safely re-use the input {@code Payload}. By contrast, {@code
+ *       RSocketClient} accepts a {@link Publisher} of {@link Payload} as input and does allow
+ *       safely re-subscribing in order to repeat a request.
+ * </ul>
+ *
+ * <p>The example below shows to create an {@code RSocketClient}:
+ *
+ * <pre class="code">{@code
+ * Mono<RSocket> source =
+ *         RSocketConnector.create()
+ *                 .metadataMimeType("message/x.rsocket.composite-metadata.v0")
+ *                 .dataMimeType("application/cbor")
+ *                 .connect(TcpClientTransport.create("localhost", 7000));
+ *
+ * RSocketClient client = RSocketClient.from(source);
+ * }</pre>
+ *
+ * <p>The below configures retry logic to use when a shared {@code RSocket} connection is obtained:
+ *
+ * <pre class="code">{@code
+ * Mono<RSocket> source =
+ *         RSocketConnector.create()
+ *                 .metadataMimeType("message/x.rsocket.composite-metadata.v0")
+ *                 .dataMimeType("application/cbor")
+ *                 .reconnect(Retry.fixedDelay(3, Duration.ofSeconds(1)))
+ *                 .toRSocketClient(TcpClientTransport.create("localhost", 7000));
```

Comment:
`toRSocketClient` should probably be something different.

Reply:
indeed